File: Level2/게임 맵 최단거리.cpp
#include <vector>
#include <iostream>
#include <queue>

using namespace std;

int solution(vector<vector<int>> maps)
{
    int answer = 1;
    int n = maps.size();
    int m = maps[0].size();
    int goalX = n - 1, goalY = m - 1;
    int directX[4] = {-1, 1, 0, 0};
    int directY[4] = {0, 0, -1, 1};
    queue<pair<int, int>> q;
    vector<vector<bool>> visited(n, vector<bool>(m, false));
    
    q.push(make_pair(0, 0));
    visited[0][0] = true;
    
    while(!q.empty()){
        int size = q.size();
        
        for(int i = 0; i < size; i++){
            int x = q.front().first;
            int y = q.front().second;
            
            q.pop();
                        
            for(int j = 0; j < 4; j++){
                int nextX = x + directX[j];
                int nextY = y + directY[j];
                
                if(nextX >= 0 && nextX < n && nextY >= 0 && nextY < m && maps[nextX][nextY] != 0 && !visited[nextX][nextY]){
                    if(nextX == goalX && nextY == goalY){
                        return answer + 1;
                    }
                    q.push(make_pair(nextX, nextY));
                    visited[nextX][nextY] = true;
                }
            }
        }
        answer++;
    }
    
    return -1;
}

File: Level2/게임 맵 최단거리.js
function solution(maps) {
    let answer = 1;
    const visited = maps;
    const queue = [];
    const directX = [-1, 1, 0, 0];
    const directY = [0, 0, -1, 1];
    const n = maps.length;
    const m = maps[0].length;
    
    queue.push([0, 0]);
    visited[0][0] = 0;
    
    while(queue.length > 0){
        const size = queue.length;
        
        for(let i = 0; i < size; i++){
            const [x, y] = queue.shift();
            
            for(let j = 0; j < 4; j++){
                const nx = x + directX[j];
                const ny = y + directY[j];

                if(nx >= 0 && nx < n && ny >= 0 && ny < m && visited[nx][ny] === 1){
                    if(nx == n - 1 && ny == m - 1){
                        return ++answer;
                    }
                    queue.push([nx, ny]);
                    visited[nx][ny] = 0;
                }
            }    
        }
        answer++;
    }
    
    return -1;
}